### PR TITLE
feature: added cross join support

### PR DESCRIPTION
**Pull request by janpaepke**
This fixes https://github.com/oguimbal/pgsql-ast-parser/issues/49 and adds support for `CROSS JOIN` as well as selecting from multiple tables (implicit cross join).

I considered parsing `tableA, tableB` to the AST equivalent of a cross join, but figured it's better to keep things symmetric and instead allowed the `from` array to contain multiple tables, which don't necessarily have to be joins.

No existing tests were modified, so backward compatibility can be assumed.

### CODE REVIEW

**Comment by janpaepke** [comment]
discovered some more edge cases and decided to tackle them. will update PR soon.

**Comment by janpaepke** [comment]
Hey Oliver,

So this took me a quite a while, as I first had to wrap my head around nearley, come up with an apparent solution (initial PR), discover a regression (PR back in DRAFT), then take the detour of arriving at a convoluted and complicated solution (which finally worked on all my test-cases), getting to a _click_ moment of actually understanding nearley and subsequently rewriting all my grammar changes to a much shorter, fairly elegant solution.

Here's the story of what happened:

### now working
Well, `CROSS JOIN`s (Duh).  
The parser can now understand `select * from ta cross join tb`.

It also understands implicit `CROSS JOIN`s, i.e. `select * from ta, tb`.

I wanted to be sure I don't cause regressions, so I did not touch any of the existing tests, but still discovered one:

My initial solution broke multiple joins, i.e. `select * from ta cross join tb cross join tc`. This had worked before – with other joins of course.

So I added some test cases and while I was at it, I thought "Well what about parantheses? Those are fun, right?".  

While the previous version did support `select * from (ta join tb [on...])` it did not support joining another table, i.e. `select * from (ta join tb [on...]) join [on...]`.

It also didn't like redundant parentheses, which are technically also valid: `select * from ((ta left join tb [on...]))`.  

Sooo long story short: All of this works now. (see test cases for details).

### (still) not working

The one test case that eluded me was `select ta cross join (tb cross join tc)`, basically the case where the joined from_item is a join itself. Again technically valid but I always ran into ambiguity issues due to the recursion.

Finally with this also not working before (with other joins) and my time spent on this already being a lot more than I had planned, I conceded and figured nobody would be too angry with me if I leave this treasure for a future adventurer.

### naming things

One of my good friends always says:
> The two hardest parts of software development are:
> 1. naming things;
> 2. cache invalidation; and
> 3. off-by-one errors.

I took the liberty of (re) naming some of the nonterminals:

- `select_subject` -> `select_from_item` it's how it's called in the [psql docs](https://www.postgresql.org/docs/9.5/sql-select.html#SQL-FROM), as there can be more than one now, we select from `select_from_items`
- `select_table_base` -> `select_from_subject` This confused me initially, as it's not _only_ a table. It's also a function or new statement. As I killed `subject` I gave it a revival.
- `select_subject_joins` -> `select_from_item_joins` for obvious reasons

### Join condition is required

While working on this and reading the psql docs for the gazillionth time, I noticed that in the grammar the join condition (ON / USING) is optional, where in reality this is only the case for `CROSS JOIN`. 

>For the `INNER` and `OUTER` join types, a join condition must be specified, namely exactly one of `NATURAL`, `ON` **join_condition**, or `USING` (**join_column** [, ...]). See below for the meaning. For `CROSS JOIN`, none of these clauses can appear.

While I thought this should be an easy fix (he wrote naively) I left this untouched, as it
a) wasn't part of this pr
b) might have been intentional to make the parser more forgiving
c) would quite surely constitute a breaking change

### Test environment

I wanted to end with a deep bow before the magnificent testing environment. It's concise, incredibly fast, implicitly checks things like ambiguity and is debugging friendly.
Without this joy I would very likely not have made it this far.

The only thing I couldn't figure out was how to run just one specific test in watch mode.
I did install the Mocha Test Explorer, which let me run single tests, but not in watch mode...
I also couldn't use `test.only` because of the helper functions.
I ended up duplicating the `select.spec.ne`, removed everything except my test and ran `npx mochapack src/syntax/select-copy.spec.ts --watch`.

I was wondering if there's a smarter way to do it?

**Comment by oguimbal** [comment]
TLDR: Congrats.

- I really did not expect someone to push something like this on the first try ! Diving into this codebase, understanding nearley, and the helper functions I used in it is definitely tricky, so well done !

I definitely agree with your good friend ... And I'll admit that I'm often lazy when naming things... not being a native english speaker doesnt help. So there is absolutely no problem renaming things to better names.


"Join condition is required" 👉 Not a problem. A lot of other incoherent statements are valid for pgsql-ast-parser. That's not the parsers' job to enforce statements functional coherency. It only is to generate an ast from a valid syntax. If you pay attention closely to pgsql error messages, you will see that syntax errors and "statement coherency errors" (I dont know how to call them) are very different, that is I guess meaning that the official internal parser will also happily output incoherent syntax trees. But that's the interpreters' job to catch those !

In my case, those use cases will be thrown by [pg-mem](https://github.com/oguimbal/pg-mem)'s execution engine.


> nobody would be too angry with me if I leave this treasure for a future adventurer.
😄 no problem

> I wanted to end with a deep bow before the magnificent testing environment.
🙏🙏🙏

> The only thing I couldn't figure out was how to run just one specific test in watch mode.

[I forked mocha test explorer](https://marketplace.visualstudio.com/items?itemName=oguimbal.vscode-mocha-test-adapter) to support HMR support, so maybe you have installed the wrong extension ?
1) Install this extension
2) `npm start`  to start the webpack build
3) reload tests
4) Optinally hit F5 to attach to the test worker (if you want to debug tests) - but debug mode is pretty useless when debugging Nearley since it does not generate sourcemaps.
4) run a test
5) tip: I've bound the `test-explorer.rerun`  command to CTRL+F5 shortcut... quite handy to re-run a test quickly when you're working on it)
<p align=center>

</p>





This has been a while now that this parser did not support cross joins... You've lifted me a great weight. I'll now have to implement it into [pg-mem](https://github.com/oguimbal/pg-mem) !

Thanks A LOT.

**Comment by oguimbal** [comment]
Released as `pgsql-ast-parser@10.0.0` 🥳

**Comment by janpaepke** [comment]
Hi Oliver,
re: testing: yeah this is what I did, as per the instructions in the [readme](https://mjam.page.link/4SF5tkM9wUFRrntcA).
It did work like you are showing in the screencast.

The `test-explorer.rerun` is a good tip, which I had not known although it's one more key hit for lazy old me.
I usually fire up a specific test in watch mode, work on stuff and want it to rerun whenever I save.

I did also realize that I have to restart the process when I broke the compilation. Like you though it didn't bother me enough. ;)

Glad I could help with this. And I learned something new along the way, which is always a good thing.
